<a href="https://colab.research.google.com/github/schaidez2727/MAT-421/blob/main/ModuleA_MAT421_Chaidez.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Module A (HW 1) - MAT 421 #
#### Santana Chaidez ####

***
## **Base-N and Binary** ##
***

In [1]:
import numpy as np
# importing Python library, NumPy, which fascilitates various forms of scientific computing

### Base-10 / Decimal System: ###
* The most commonly used numeric system, and the decimal system that we all learn, is a **base-10** system that uses the digits 0-9 as coefficients for various powers of 10
* This understanding of the base-10 system can be used to also understand any other **base-N** system, where each number in the system is represented by the list of digits 0-(N-1) and each digit represents the coefficient for a power of N

In [2]:
# Decimal expansion of 1234.56 (base-10)
1234.56 == 1*10**(3) + 2*10**(2) + 3*10**(1) + 4*10**(0) + 5*10**(-1) + 6*10**(-2)
# ^^ Shows each digit of 1234.56 (base-10) as a coefficient of a descending power of 10

True

### Misc. Base-N Decompositions into Base-10 ###

In [3]:
# 102 (base-3) = 11 (base-10)
1*3**(2) + 0*3**(1) + 2*3**(0) == 11

True

In [4]:
# 13 (base-13) = 16 (base-10)
1*13**(1) + 3*13**(0) == 16

True

### Base-2 / Binary: ###
* **Binary**, the representation of numbers commonly used by computers, is actually a **base-2** system
* The system represents numbers using the digits 0 and 1, which each represent the coefficient for a power of 2
* Computers benefit from using only digits 0 and 1 since they can also act as boolean/true-false indicators, and additionally can have arithmetic operations represented using AND, OR, and NOT
* Faster and more efficient for computers in many ways, but limits amount of numbers able to be represented depending on amount of bits the computer has

### Base-10 to Binary Conversion ###

In [5]:
# First want to separate the base-10 number into a sum of powers of 2
# e.g. 2^0=1, 2^1=2, 2^2=4, 2^3=8, 2^4=16, 2^5=32, 2^6=64, etc.

#Decomposition of 27(base-10) into powers of 2:
27 == 16 + 8 + 2 + 1

True

In [6]:
# Next, want to identify the higest power of 2 used and identify the coeffcient for each consecutive power in descending order
# For 27 (base-10), the highest power of 2 it contains is 4 --> 2^4 = 16

#Rewrite sum in powers of 2 with a coefficient of either 0 or 1:
27 == 1*2**(4) + 1*2**(3) + 0*2**(2) + 1*2**(1) + 1*2**(0)

True

In [7]:
# Finally, combine the coefficients to identify the binary/base-2 conversion of 27(base-10)
print('27(base-10) = 11011(base-2)')

# Check with NumPy Decimal to Binary representation operator
'11011' == np.binary_repr(27)

27(base-10) = 11011(base-2)


True

In [8]:
# Let's recap with one more example:
# 13(base-10) = 1101(base-2)

# Assessing accuracy of base-10 decomposition into base-2
print(13 == 8 + 4 + 1)
print(13 == 1*2**(3) + 1*2**(2) + 0*2**(1) + 1*2**(0))

# Identifying and verifying base-2 conversion
print('13(base-10) = 1101(base-2)')
'1101' == np.binary_repr(13)

True
True
13(base-10) = 1101(base-2)


True

### Addition and Multiplication: Binary and Decimal Systems ###
Base-2 Arithmetic Remunder:
1 1 0 1 **1**  +  1 1 0 **1**  -->  1 + 1 = 10 --> We would put a 0 in this first place within the sum and "carry the one" to the next highest place as we would for a sum of 10 or more in decimal addition!

In [9]:
# Base-10: 27 + 13 = 40
# Base-2: 11011 + 1101 = 101000

# Base-10 decomposition into base-2
print(40 == 1*2**(5) + 0*2**(4) + 1*2**(3) + 0*2**(2) + 0*2**(1) + 0*2**(0))

# Check for accuracy:
print('40(base-10) = 101000(base-2)')
'101000' == np.binary_repr(40)

True
40(base-10) = 101000(base-2)


True

In [10]:
# Base-10: 27 * 13 = 351
# Base-2: 11011 * 1101 = 101011111

# Base-10 decomposition into base-2
print(351 == 1*2**(8) + 0*2**(7) + 1*2**(6) + 0*2**(5) + 1*2**(4) + 1*2**(3) + 1*2**(2) + 1*2**(1) + 1*2**(0))

# Check for accuracy:
print('351(base-10) = 101011111(base-2)')
'101011111' == np.binary_repr(351)

True
351(base-10) = 101011111(base-2)


True

***
## **Floating Point Numbers** ##
***

* Floating point numbers, or float, are how we expand the range of values computers can use in operations and calculations with the same amount of bits
* Expands the range beyond just positive integers to account for decimals and negative values
* Floats are commonly mapped using the **IEEE754** system. For 64-bit systems, this means 1 bit/digit is the sign indicator **s**, the next 11 bits determine the chracteristic/exponent **e**, and the final 52 bits determine the fraction **f** in the following format:
* floating point n = (-1)^s x 2^(e-1023) x (1+f)

In [11]:
# IEEE754 to Decimal Conversion
# 1 10000000001 1010000000000000000000000000000000000000000000000000 (IEEE754) = -6.5 (base-10)

# Idenitfy s, e, and f:
s_1 = 1
e_1 = 1*2**(10) + 1*2**(0)
f_1 = 1/(2**1) + 0/(2**2) + 1/(2**3)

# Plug into IEEE754 float representation:
n_1 = (-1)**s_1 * 2**(e_1-1023) * (1+f_1)
print('1 10000000001 1010000000000000000000000000000000000000000000000000 (IEEE754) =')
print(n_1)

1 10000000001 1010000000000000000000000000000000000000000000000000 (IEEE754) =
-6.5


* Due to the fraction component of the floating point formula, there are gaps between each value that can be computed using this system
* And due to the exponential in this formuala, the gap between values gets larger with the values being computed
* Values within these gaps are assigned to their nearest "neighbor" value

In [12]:
# Computing the gap between neighboring values at given large number, 1e7
gap = np.spacing(1e7)
print(gap)

1.862645149230957e-09


In [13]:
# Verifying gap by adding less than half of the gap value back to 1e7
1e7 == 1e7 + (gap/2.5)

True

* Due to system limitations dependent on the amount of bits available, there is still a smallest and largest decimal value that a computer can recognize in its computations
* Numbers greater than the largest representable floating point number for a system result in **overflow**, where the result is aassigned to the value of infinity (**inf**)
* Number less than the smallest representable float result in **underflow**, where the assigned result is **0**

In [14]:
import sys
sys.float_info
# Displays float information and specifications for our system

sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)

The previous output identifies the **largest** representable floating point number as `sys.float_info.max = 1.7976931348623157e+308` amd the **smallest** representable floating point number as `sys.float_info.min = 2.2250738585072014e-308`

In [15]:
# Displaying overflow
2 * sys.float_info.max

inf

In [16]:
# Displaying underflow
2**(-500) * sys.float_info.min

0.0

***
## **Round-Off Errors** ##
***

* A result of using base-2 fractions to represent floating point numbers is that the numbers must be approximated according to the finite number of bytes used by the computer
* This results in a **round-off error**, which is the difference between the computer's approximation of the number and it's true value
* This approximation causes small errors in arithmetic operations

In [17]:
# As an example, let's start with a small difference between 2 decimal numbers
3.7 - 3.665

0.03500000000000014

In [18]:
# However, we know the difference should just be 0.035, but the system will not agree
3.7 - 3.665 == 0.035

False

* This is the result of the round-off error that occurs during the operation due to the floating point numbers being represented by approximations!
* We can observe more comparable results by rounding

In [19]:
print(0.1 + 0.2 + 0.4 == 0.7)
# Despite us knowing this is accurate, the output will read False due to error

round(0.1 + 0.2 + 0.4, 5) == round(0.7, 5)
# But rounding both operations will make them equivalent to the computer!

False


True

* This round-off error can accumulate and be amplified by repeated calculations

In [20]:
# Let's take, for example, the operation 1 + 1/3 - 1/3
1 + (1/3) - (1/3)
# Done just once, the operation correctly produces a result of 1

1.0

In [21]:
# Now, let's try repeating this operation with an iterative function defined using for loops
def add_sub(iterations):
    result = 1
    # setting the initial (correct) result

    for i in range(iterations):
        result += 1/3
        # adds 1/3 to result of previous iteration

    for i in range(iterations):
        result -= 1/3
        # subtracts 1/3 from result pf previous iteration

    return result

In [22]:
# We can observe how the round-off error accumulates by running this operation for an increasing number of iterations

print(add_sub(100))
# repeatedly adding and subtracting 1/3 from 1 100 times

print(add_sub(1000))
# repeatedly adding and subtracting 1/3 from 1 1000 times

print(add_sub(10000))
# repeatedly adding and subtracting 1/3 from 1 10000 times

print(add_sub(1000000))
# repeatedly adding and subtracting 1/3 from 1 one million times

1.0000000000000002
1.0000000000000064
1.0000000000001166
0.9999999999727986


The variation from the actual value (1) increases with the number of iterations!